### PR TITLE
Add a "KeyAlreadyRegistered" exception for Client.register [needs revision]

### PR DESCRIPTION
**Pull request by wteiken**
Add a specific KeyAlreadyRegistered exception that contains the previous registration URL.

This includes adding the response object to the ClientError exception.  It also includes returning the mesages.Error parsed from a non-success response in a field of ClientError instead of as a separate exception.

### CODE REVIEW

**Comment by pde** [comment]
Integration tests failed due to a heisenbug we'd been wrestling with; restarting them.

**Comment by wteiken** [comment]
There are also pieces missing to adopt to this in the rest of the client (although no integration test seems to fail on my test side).   I am not sure if this is the right way to go, somebody who has a better idea where the API is supposed to head should chime in...

My motivation is that while it is nice to have the parsed error object it loses a lot of the context and due to the nature of message.Error it can't easily be added there.  So it seemed natural to make the message.Error a field of the ClientError and uniformly return that (and probably also no longer make message.Error a subclass of Exception as the ClientError will be caught).  It may be nicer to throw a subclass of ClientError (and not have to check if error is set).  I am pretty much already planning to do that as it will simplify code.

There are a few references to message.Error in auth_handler that I am not sure yet how they are handled best...  so this pull is more an attempt to start the discussion, not a final solution.

There are a few sites in the client where the message.Error

**Comment by kuba** [comment]
This PR likely fixes #829 but stands in conflict with my local changes... It would be better to sync before duplicating each other's work. Will attempt to review some time this week.

File: acme/acme/errors.py [hunk 1]
```diff
@@ -11,26 +11,62 @@ class SchemaValidationError(jose_errors.DeserializationError):
 
 
 class ClientError(Error):
+    """Network error or unexpected client response."""
+    def __init__(self, response, *args, **kwargs):
+        """
+        :param requests.Response details: The response object that caused the
```
Comment (kuba) [inline acme/acme/errors.py]: Move docstring to `ClientError.__doc__` (using `ivar`), rename `details` to `response`

**Comment by pde** [comment]
@wteiken I believe the main places to look at where your exceptions could be caught are `letsencrypt/cli.py` (especially `_handle_exception` and `_determine_account`) and `letsencrypt/client.py` (especially `perform_registration`).

**Comment by siccegge** [comment]
Hi! what's the state of this?

**Comment by bmw** [comment]
@siccegge, it appears Kuba left some comments on the PR that were never addressed and there are merge conflicts that need to be fixed. If these changes are made, I will review.

**Comment by jsha** [comment]
I have a narrower fix for this, introducing a ConflictError in #4254. Since this PR is relatively old and has a lot of conflicts, I'm closing it for now.

**Pull request closed without merging**